storage: replace the rule solver

BramGruneir opened the pull request:
The 1st commit is a collection of fixes and updates to the rule solver.

The 2nd commit adds back in the randomness that was missing from the rule solver.

The 3rd commit removes the rule solver altogether and replaces it with a collection of functions for the rules and 3 functions for allocation, removal and rebalancing.

There's more work to do here after merging, but this will bring us close enough to start testing rebalancing using this more expressive system.

With this PR #11702 and #11721 can be closed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12165)
<!-- Reviewable:end -->

BramGruneir committed:
storage: collection of fixes and updates for the rules solver

- Splits the scores returned in the rule solver into a constraint and balance scores.
- Add a valid field to constraints and add it to all rules.
- Solve now returns all candidates instead of just the valid ones. To get only the valid candidates, the new function onlyValid and new type  condidateList have also been added.
- This allows us to use solve for removeTarget. It also cleans up the logic in removeTarget to more closely match the non-rule solver version.
- Split the capcity rules into two rules. They were performing two different operations and didnt' make sense being combined. This will also ease the change of converting the rules to basic functions.

Part of #10275

BramGruneir committed:
storage: add a little randomness into the rule_solver

This commit adds the equivalent of the current selectGood and selectBad so that the rule solver will also use the "power of two random choices" method that is currently used by the balancer.

BramGruneir committed:
storage: replace the rule solver

Remove the concept of rules and replace it with clean direct functions.

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 362 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUEFkkRKFzuD9DP_WC:-KYUEFkkRKFzuD9DP_WD:b-s16lm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L362)):*
> ```Go
> 	if a.options.UseRuleSolver {
> 		// TODO(bram): ShouldRebalance should be part of rebalanceCandidates
> 		// and decision made afterward, not it's own function.
> ```

You probably have a reason in mind. Add it to the comment.

---

*[pkg/storage/allocator.go, line 392 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUPMnk_F5yNcP8FhZz:-KYUPMnk_F5yNcP8Fh_-:byi2bgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L392)):*
> ```Go
> 
> 		// Find all candidates that are better than the worst existing replica.
> 		targets := candidates.betterThan(existingCandidates[len(existingCandidates)-1])
> ```

This will crash if `len(existingCandidates) == 0` for some reason.

---

*[pkg/storage/rule_solver.go, line 33 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNo9urH7Bd5KdTp5m:-KYUNo9urH7Bd5KdTp5n:b-3pxbzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L33)):*
> ```Go
> 	valid      bool
> 	constraint float64 // Score used to pick the top candidates.
> 	balance    float64 // Score used to choose between top candidates.
> ```

`s/balance/capacity/g` to reflect that this is always initialized to `capacityScore`.

---

*[pkg/storage/rule_solver.go, line 37 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUIGl491B54v6sFBAm:-KYUIGl491B54v6sFBAn:bkbutla) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L37)):*
> ```Go
> 
> func (c candidate) String() string {
> 	return fmt.Sprintf("StoreID:%d, valid:%t, con:%.2f, bal:%.2f",
> ```

`s/StoreID:%d/s%d` to match the format used elsewhere. 

---

*[pkg/storage/rule_solver.go, line 96 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKw6bjGpFb8JfgXAH:-KYUKw6bjGpFb8JfgXAI:b-5wpkkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L96)):*
> ```Go
> func (c byScoreAndID) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
> 
> // onlyValid returns all the elements in a sorted candidate list that are valid.
> ```

You should add a comment to all of these `candidateList` methods that they require the candidates to be sorted (and which sorting is required).

---

*[pkg/storage/rule_solver.go, line 115 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKZd82aPgqrxuy4O9:-KYUKZd82aPgqrxuy4OA:b-m08ooq) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L115)):*
> ```Go
> 	for i := 1; i < len(cl); i++ {
> 		if cl[i].constraint < cl[0].constraint {
> 			return cl[0:i]
> ```

`s/0:i/i/g`

---

*[pkg/storage/rule_solver.go, line 127 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKhK38YzcKgVz7XV3:-KYUKhK38YzcKgVz7XV4:b-k3aula) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L127)):*
> ```Go
> 		return cl
> 	}
> 	// Are there invalid values? If so, pick those.
> ```

`s/values/candidates/g`

---

*[pkg/storage/rule_solver.go, line 156 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUCYHXZ86MTv3JBSGp:-KYUCYHXZ86MTv3JBSGq:b-aqvrvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L156)):*
> ```Go
> // selectGood randomly chooses a good candidate from a sorted candidate list
> // using the provided random generator.
> func (cl candidateList) selectGood(randGen allocatorRand) candidate {
> ```

Return a `*candidate` so that you don't require the caller to perform an allocation.

---

*[pkg/storage/rule_solver.go, line 164 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYULCqW02aDFJNeD5fv:-KYULCqW02aDFJNeD5fw:b8m4yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L164)):*
> ```Go
> 	order := randGen.Perm(len(cl))
> 	randGen.Unlock()
> 	best := cl[order[0]]
> ```

This will crash if `len(cl) == 0`.

---

*[pkg/storage/rule_solver.go, line 183 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYULGt7Rt-qggHvDpMb:-KYULGt7Rt-qggHvDpMc:b-zfog90) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L183)):*
> ```Go
> 	order := randGen.Perm(len(cl))
> 	randGen.Unlock()
> 	worst := cl[order[0]]
> ```

Ditto crash mentioned above.

---

*[pkg/storage/rule_solver.go, line 234 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUMahvHqXWAWV34O45:-KYUMahvHqXWAWV34O46:b-idkag2) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L234)):*
> ```Go
> // removeCandidates creates a candidate list of all existing replicas' stores
> // ordered from least qualified for removal to most qualified. Stores that are
> // marked as not valid, are in violation of a criteria.
> ```

`s/criteria/required criteria/g`?

---

*[pkg/storage/rule_solver.go, line 254 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUO-ptC9QCJMz9D-W-:-KYUO-ptC9QCJMz9D-W0:b-qxwbq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L254)):*
> ```Go
> 		constraintScore := diversityRemovalScore(s.Node.NodeID, existingNodeLocalities) + float64(preferredMatched)
> 		if rebalanceToConvergesOnMean(sl, s) {
> 			constraintScore++
> ```

What does it mean for `rebalanceToConvergesOnMean` to affect `constraintScore`?

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYUNWabK8aXFckYbZeJ:bxwy4mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
> ```Go
> // the highest scores.
> func capacityScore(store roachpb.StoreDescriptor) float64 {
> 	return 1.0 / float64(store.Capacity.RangeCount+1)
> ```

Does this need to return a value between 0 and 1 now? You're randomly selecting from the candidates with the same constraint score and then only comparing this value against itself.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator.go, line 362 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUEFkkRKFzuD9DP_WC:-KYVKJtBsGXusBgkatTm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L362)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You probably have a reason in mind. Add it to the comment.
</blockquote></details>

Done.

---

*[pkg/storage/allocator.go, line 392 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUPMnk_F5yNcP8FhZz:-KYVKc5sd8S16kvqE4Ch:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L392)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This will crash if `len(existingCandidates) == 0` for some reason.
</blockquote></details>

fixed

---

*[pkg/storage/rule_solver.go, line 33 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNo9urH7Bd5KdTp5m:-KYVL9NC4GtFipbSYujO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L33)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/balance/capacity/g` to reflect that this is always initialized to `capacityScore`.
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 37 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUIGl491B54v6sFBAm:-KYVLsJjGfFlKIaOVHcR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L37)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/StoreID:%d/s%d` to match the format used elsewhere. 
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 96 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKw6bjGpFb8JfgXAH:-KYVMC4J2DyDRD3-ZS8G:b-hv4j7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L96)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You should add a comment to all of these `candidateList` methods that they require the candidates to be sorted (and which sorting is required).
</blockquote></details>

They already said "sorted", but I've added a "(by score reversed)" to them.

---

*[pkg/storage/rule_solver.go, line 115 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKZd82aPgqrxuy4O9:-KYVNBs-ck2PNTRtK1-m:bypefhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L115)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/0:i/i/g`
</blockquote></details>

you meant `:i` right?  If so, done.

---

*[pkg/storage/rule_solver.go, line 127 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUKhK38YzcKgVz7XV3:-KYVNLr1okvPDmVOhLGx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L127)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/values/candidates/g`
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 156 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUCYHXZ86MTv3JBSGp:-KYVNUAhp8_fDr3X_cZg:b-s8329l) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L156)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Return a `*candidate` so that you don't require the caller to perform an allocation.
</blockquote></details>

Done. Actually, changed this and selectBad to return a pointer to a storeDescriptor.

---

*[pkg/storage/rule_solver.go, line 164 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYULCqW02aDFJNeD5fv:-KYVP-Q6xRESLnb5o5z_:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L164)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This will crash if `len(cl) == 0`.
</blockquote></details>

fixed

---

*[pkg/storage/rule_solver.go, line 183 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYULGt7Rt-qggHvDpMb:-KYVP2S_NGT-Umf-5DOI:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L183)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ditto crash mentioned above.
</blockquote></details>

fixed

---

*[pkg/storage/rule_solver.go, line 234 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUMahvHqXWAWV34O45:-KYVPBU4XcqA7VNbsx8q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L234)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/criteria/required criteria/g`?
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 254 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUO-ptC9QCJMz9D-W-:-KYVPDRTjj8-f-d0XjuJ:bdl3i62) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L254)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What does it mean for `rebalanceToConvergesOnMean` to affect `constraintScore`?
</blockquote></details>

If we consider that we want to converge on toward mean when making allocation decisions, than it is a constraint (just not a formal constraint from our zone configs).

This allows us to use the criteria from shouldRebalance as part of ordering the candidates to make this decision in one place with full knowlege. 

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYVPK2CjRQyMcvRrbbV:bhg61yx) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does this need to return a value between 0 and 1 now? You're randomly selecting from the candidates with the same constraint score and then only comparing this value against itself.
</blockquote></details>

It doesn't have to be between 0 and 1, but there's no reason not no.  We want it to be higher for more empty replicas.  It would be strange to order by constraint score descending and capacity score ascending. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator.go, line 392 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUPMnk_F5yNcP8FhZz:-KYZAp1tPd7zGMSuSuTA:bowvtwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L392)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

I might be missing the fix, but it still looks like this will crash if `len(existingCandidates) == 0`. 

---

*[pkg/storage/rule_solver.go, line 254 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUO-ptC9QCJMz9D-W-:-KYZA39Nzl7vVhvLuYVv:buhd2bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L254)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

If we consider that we want to converge on toward mean when making allocation decisions, than it is a constraint (just not a formal constraint from our zone configs).

This allows us to use the criteria from shouldRebalance as part of ordering the candidates to make this decision in one place with full knowlege. 
</blockquote></details>

Should this be `rebalanceFromConvergesOnMean`?

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYZA8KlKwHuzHssPWmG:b-z220ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It doesn't have to be between 0 and 1, but there's no reason not no.  We want it to be higher for more empty replicas.  It would be strange to order by constraint score descending and capacity score ascending. 
</blockquote></details>

It's also a little strange to be forcing the score into the range 0 to 1. And in order to do that you need to use `RangeCount+1`. I don't particularly care. Pick your strangeness.

---

*[pkg/storage/rule_solver.go, line 169 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYZ9VHe2vGRspxFZ-HR:-KYZ9VHe2vGRspxFZ-HS:b-xje8fo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c773d3c2a774cc0621f60ad2620e4addac57d1/pkg/storage/rule_solver.go#L169)):*
> ```Go
> 	order := randGen.Perm(len(cl))
> 	randGen.Unlock()
> 	best := cl[order[0]]
> ```

`s/cl[order[0]]/&cl[order[0]]/g` so that `return &best.store` does not cause an unnecessary allocation.

---

*[pkg/storage/rule_solver.go, line 191 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYZ9eDVA4Ggoq-vottl:-KYZ9eDVA4Ggoq-vottm:bxq96jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c773d3c2a774cc0621f60ad2620e4addac57d1/pkg/storage/rule_solver.go#L191)):*
> ```Go
> 	order := randGen.Perm(len(cl))
> 	randGen.Unlock()
> 	worst := cl[order[0]]
> ```

See above comment about using a pointer for `worst`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator.go, line 392 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUPMnk_F5yNcP8FhZz:-KYlLapv52FirWgVGa6V:b-48jgy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/allocator.go#L392)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I might be missing the fix, but it still looks like this will crash if `len(existingCandidates) == 0`. 
</blockquote></details>

Ah, I fixed a different one. 
I've put in a fix for this, but it shouldn't happen.  This requires calling rebalance on a replica in which no store is alive in the store pool, including this one.  
Perhaps a panic here is a good idea.

---

*[pkg/storage/rule_solver.go, line 254 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUO-ptC9QCJMz9D-W-:-KYgQy4V1txXcu5UAHqi:b3bwr3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L254)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be `rebalanceFromConvergesOnMean`?
</blockquote></details>

We actually want the opposite.   SelectBad in the balancer calls that and uses it to identify preferred candidates for removal.  In this case, we want the worse candidates to have a lower score.  

I considered removing 1 form the constraint score, but I was worried about the interaction it might have with positive zone config constraints or diversity scores, both of which are always positive.

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYgRZD6weaJlPujxTe0:b-zeipcc) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It's also a little strange to be forcing the score into the range 0 to 1. And in order to do that you need to use `RangeCount+1`. I don't particularly care. Pick your strangeness.
</blockquote></details>

I'd be down for changing this up (and the already strange sort reverseness), but in a follow up PR.  If we do come up with a collection of metrics to determine sizes, I'd rather make them range from -1 to 1 which more closely matches ML style scores.

---

*[pkg/storage/rule_solver.go, line 169 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYZ9VHe2vGRspxFZ-HR:-KYgO70Jav-wCIt6_uj2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c773d3c2a774cc0621f60ad2620e4addac57d1/pkg/storage/rule_solver.go#L169)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/cl[order[0]]/&cl[order[0]]/g` so that `return &best.store` does not cause an unnecessary allocation.
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 191 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYZ9eDVA4Ggoq-vottl:-KYgO7imNt4vmSHQE7-x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c773d3c2a774cc0621f60ad2620e4addac57d1/pkg/storage/rule_solver.go#L191)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

See above comment about using a pointer for `worst`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYnKkIx1LAx-W2QdE2N:b-o84hsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
Ok on adjusting this in a follow-on PR.

> which more closely matches ML style scores
 
Why is that important?

---

*[pkg/storage/rule_solver.go, line 260 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYnLuhkCBEQrry7A2YD:-KYnLuhkCBEQrry7A2YE:b5vnj5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51aaefd403b7aaa46df10b1b80a77bdf4f36db6/pkg/storage/rule_solver.go#L260)):*
> ```Go
> 		}
> 		constraintScore := diversityRemovalScore(s.Node.NodeID, existingNodeLocalities) + float64(preferredMatched)
> 		if rebalanceToConvergesOnMean(sl, s) {
> ```

Ah, so should this be `rebalanceFromConvergesOnMean`. It doesn't make sense to be using the same function in both `removeCandidates` and `rebalanceCandidates`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYnyI9lGwDBvrFuubs9:bh82xeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok on adjusting this in a follow-on PR.

> which more closely matches ML style scores
 
Why is that important?
</blockquote></details>

It's not, but since it's more or less a standard, we might be able to leverage some basic ML techniques if we have a number of criteria in need of weighting.  
Note that this is not going to be needed anytime soon and hand tuned heuristics should be fine for a long time.

I'll send a follow up PR that cleans this up.  

---

*[pkg/storage/rule_solver.go, line 260 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYnLuhkCBEQrry7A2YD:-KYoXyP-zKhMLUwVOzIM:b-cmhm8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51aaefd403b7aaa46df10b1b80a77bdf4f36db6/pkg/storage/rule_solver.go#L260)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ah, so should this be `rebalanceFromConvergesOnMean`. It doesn't make sense to be using the same function in both `removeCandidates` and `rebalanceCandidates`. 
</blockquote></details>

Ok, so I had to step away from this and re-examine exactly how this should work.  I think I've got it.  Please let me know if my logic is flawed here.

First of all, `rebalanceFromConvergesOnMean !=  !rebalanceToConvergesOnMean` which was an assumption I had made.

With that in mind:

Removals:  What we actually want is `!rebalanceFromConvergesOnMean` as it matches the selectBad by promoting those that do converge on the mean.

Rebalance: For the existing candidates, we want `!rebalanceFromConvergesOnMean` exactly like the removals. While for new candidates, we want only stores that `rebalanceToConvergesOnMean` Which matches `improve'.  This keeps the operations in line with the original rebalancing.  So for all candidates, we eliminate any that don't `rebalanceToConvergeOnMean` and add 1.0 to the base constraint score to match the positive removal score.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->

petermattis commented:
As you probably realize, this PR is a beast which makes it very likely something has slipped through. I don't see a good way to mitigate for this PR, but in the future smaller PRs are better: easier on the reviewer and better reviewed code for the author.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/rule_solver.go, line 439 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYUNWabK8aXFckYbZeI:-KYooIPVXxNJlaD2cjzc:b-6qhjo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/324f16fd976e4d4fcb47446c83768c5109df5475/pkg/storage/rule_solver.go#L439)):*
> we might be able to leverage some basic ML techniques if we have a number of criteria in need of weighting

That's way too speculative to use as a basis for a change. When we do have a need to weight a number of criteria we should revisit, but not before. 

---

*[pkg/storage/rule_solver.go, line 260 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYor9CiMvMmsYAxwnNv:-KYor9CiMvMmsYAxwnNw:bnkigbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L260)):*
> ```Go
> 		}
> 		constraintScore := diversityRemovalScore(s.Node.NodeID, existingNodeLocalities) + float64(preferredMatched)
> 		if !rebalanceFromConvergesOnMean(sl, s) {
> ```

This is somewhat confusing and deserves a comment: `Removing a replica from the candidate converges on the mean which makes the candidate more attractive, blah blah blah regarding selecting worst candidate`. 

---

*[pkg/storage/rule_solver.go, line 312 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYos3E2XVdDCvVDAeEe:-KYos3E2XVdDCvVDAeEf:biannpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L312)):*
> ```Go
> 			}
> 			constraintScore := diversityRemovalScore(s.Node.NodeID, existingNodeLocalities) + float64(preferredMatched)
> 			if !rebalanceFromConvergesOnMean(sl, s) {
> ```

Ditto my other comment. This one can be a reference to the comment in `removeCandidate`.

---

*[pkg/storage/rule_solver.go, line 325 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYosZJD4Uih6AQgLyWI:-KYosZJErnXwhrfeMTYM:b-cccb9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L325)):*
> ```Go
> 				continue
> 			}
> 			constraintScore := 1.0 + diversityScore(s, existingNodeLocalities) + float64(preferredMatched)
> ```

Adding a constant `1.0` won't affect the ordering of candidates. Either explain what it is doing there or remove.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-:-KYoslnhbLQFzIzn_zHD:bk17w68)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/rule_solver.go, line 260 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYor9CiMvMmsYAxwnNv:-KYp0CMnPT6hjlRXMdxJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L260)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is somewhat confusing and deserves a comment: `Removing a replica from the candidate converges on the mean which makes the candidate more attractive, blah blah blah regarding selecting worst candidate`. 
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 312 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYos3E2XVdDCvVDAeEe:-KYp5Mc1zmSZKeQ4htmP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L312)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ditto my other comment. This one can be a reference to the comment in `removeCandidate`.
</blockquote></details>

Done.

---

*[pkg/storage/rule_solver.go, line 325 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165#-KYosZJD4Uih6AQgLyWI:-KYozfIU8yQJ4Zu7jEEL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b19ec3588080596e4e6935bac70ec62e41a9c61a/pkg/storage/rule_solver.go#L325)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Adding a constant `1.0` won't affect the ordering of candidates. Either explain what it is doing there or remove.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12165)*
<!-- Sent from Reviewable.io -->
